### PR TITLE
Bluetooth: Mesh: Populate Bridge Config Server

### PR DESCRIPTION
**Pull request by omkar3141**
Adds implementation for the Bridge Configuration Server model.

Updates `brg_cfg` module to add sanity check for
bt_mesh_brg_cfg_tbl_remove() API. Also, updates the unit test accordingly.

Adds documentation for the Bridge Configuration Server model

### CODE REVIEW

**Comment by alxelax** [comment]
Please keep @PavelVPV in track since he is code owner.

File: doc/connectivity/bluetooth/api/mesh/brg_cfg_srv.rst
Comment (mia-ko) [inline doc/connectivity/bluetooth/api/mesh/brg_cfg_srv.rst:13]:
```suggestion
The Bridge Configuration Server model relies on a :ref:`bluetooth_mesh_models_brg_cfg_cli` to
```

File: doc/connectivity/bluetooth/api/mesh/brg_cfg_srv.rst
Comment (mia-ko) [inline doc/connectivity/bluetooth/api/mesh/brg_cfg_srv.rst:19]:
```suggestion
The Bridge Configuration Server model provides access to the following three states:

* Subnet Bridge state
* Bridging Table state
* Bridging Table Size state
```

Comment (mia-ko) [reply doc/connectivity/bluetooth/api/mesh/brg_cfg_srv.rst:19]: Alternatively, the sections under can be replaced by a bullet list or a definition list :)
I don't have any strong preferences, so I will leave it up to you :)

File: doc/connectivity/bluetooth/api/mesh/brg_cfg_srv.rst
Comment (mia-ko) [inline doc/connectivity/bluetooth/api/mesh/brg_cfg_srv.rst:26]:
```suggestion
This state enables or disables the Subnet Bridge feature on a node. When the Subnet Bridge feature is
enabled on a node, the node can transfer received messages from one subnet to another as
specified by the Bridging Table state.
```

File: doc/connectivity/bluetooth/api/mesh/brg_cfg_srv.rst
Comment (mia-ko) [inline doc/connectivity/bluetooth/api/mesh/brg_cfg_srv.rst:34]:
```suggestion
This state holds the bridging table. This table is used to check if incoming messages can be bridged
from one subnet to another. The maximum number of rows supported by the Bridging Table state
can be configured using the :kconfig:option:`CONFIG_BT_MESH_BRG_TABLE_ITEMS_MAX`
configuration option.
```

File: doc/connectivity/bluetooth/api/mesh/brg_cfg_srv.rst
Comment (mia-ko) [inline doc/connectivity/bluetooth/api/mesh/brg_cfg_srv.rst:40]:
```suggestion
read-only state.
```

**Review by mia-ko — CHANGES_REQUESTED** [review]

File: doc/connectivity/bluetooth/api/mesh/brg_cfg_srv.rst
Comment (mia-ko) [inline doc/connectivity/bluetooth/api/mesh/brg_cfg_srv.rst:27]:
```suggestion
  by the Bridging Table state can be configured using the
```

**Review by mia-ko — CHANGES_REQUESTED** [review]
Apart from one tiny change suggestion, looks good to me :)

File: subsys/bluetooth/mesh/brg_cfg_srv.c
Comment (HaavardRei) [inline subsys/bluetooth/mesh/brg_cfg_srv.c:301]:
```suggestion
	{ OP_BRIDGED_SUBNETS_GET, BT_MESH_LEN_EXACT(3), bridged_subnets_get },
```

**Review by HaavardRei — CHANGES_REQUESTED** [review]

File: doc/connectivity/bluetooth/api/mesh/brg_cfg_srv.rst
Comment (omkar3141) [reply doc/connectivity/bluetooth/api/mesh/brg_cfg_srv.rst:6]: We don't mention this for other models, but may be let's start from here.

**Review by alxelax — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/mesh/brg_cfg_srv.c
Comment (omkar3141) [reply subsys/bluetooth/mesh/brg_cfg_srv.c:180]: I don't think so. Please check highlighted text from below:

> The Start_Index field determines the offset in units of pairs of NetKey Indexes (see Table 4.293) to start from when **reporting the filtered set of pairs** of NetKey Indexes extracted from the Bridging Table state entries of a Subnet Bridge. 

The counting happens on the filtered set of pairs not on the raw table rows. There is also test case that checks this. Note that `start_idx` is uint8_t value here. But the actual table can be twice or thrice bigger than this.

Also see this excerpt from spec which describes how to compost the response list message:

> The Bridged_Subnets_List field shall contain a filtered set of not repeated (i.e., unique) entries that are formatted as described in Table 4.293. The filtered set includes (NetKeyIndex1, NetKeyIndex2) pairs extracted from the Bridging Table state entries that meet the filtering condition in the Filter field (see Table 4.291) and starts with the filtered (NetKeyIndex1, NetKeyIndex2) pair indicated by the Start_Index field, considering zero-based indexing (**i.e., the first filtered NetKey Indexes pair is assigned the index 0**).

File: subsys/bluetooth/mesh/brg_cfg_srv.c
Comment (omkar3141) [reply subsys/bluetooth/mesh/brg_cfg_srv.c:252]: No. There is also an [erratum ](https://bluetooth.atlassian.net/browse/ES-25822) on this to clarify wrong description in the message tables. Let's discuss IRL if you are still wondering about this.

File: subsys/bluetooth/mesh/brg_cfg_srv.c
Comment (omkar3141) [inline subsys/bluetooth/mesh/brg_cfg_srv.c:169]: TODO: Cover the clause of reported entries being **unique**:

> The Bridged_Subnets_List field shall contain a filtered set of not **repeated (i.e., unique)** entries that are formatted as described in Table 4.293.

**Review by mia-ko — APPROVED** [review]

File: subsys/bluetooth/mesh/brg_cfg_srv.c
Comment (alxelax) [reply subsys/bluetooth/mesh/brg_cfg_srv.c:22]: Screenshot from 2024-08-06 08-49-02
In most of C compilers and versions it will work. However, as you can see, C specification doesn't guarantee to 100% that `true` is always 1. There are compilers with obsolete behavior.
Code that you implemented depends on presence or absence of obsolete compiler behavior. The source code that I suggest is **really** cross-platform and compiler independent.

File: subsys/bluetooth/mesh/brg_cfg_srv.c
Comment (alxelax) [reply subsys/bluetooth/mesh/brg_cfg_srv.c:110]: I suggest to set 0 only direction. The rest of the memory is written twice that is inefficiency.

File: subsys/bluetooth/mesh/brg_cfg_srv.c
Comment (alxelax) [reply subsys/bluetooth/mesh/brg_cfg_srv.c:42]: Sorry, I didn't check it is enumeration not boolean. Nevertheless, it is not required to cast enumeration to `uint8_t`. Let compiler operate with int.

File: subsys/bluetooth/mesh/brg_cfg_srv.c
Comment (omkar3141) [reply subsys/bluetooth/mesh/brg_cfg_srv.c:22]: Thanks a lot for this.

File: subsys/bluetooth/mesh/brg_cfg.h
Comment (alxelax) [inline subsys/bluetooth/mesh/brg_cfg.h:12]: Definitions should have prefix `BT_MESH_BRG_CFG_`

**Review by alxelax — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/mesh/brg_cfg.c
Comment (omkar3141) [reply subsys/bluetooth/mesh/brg_cfg.c:314]: No, this will not work. Because then `store` will have to be declared first, and then perhaps we hope that compiler removes it. But side effect will be `store = true` few lines above will have to be kept, which will consume extra code. So I don't think it is worth simplifying this.

**Review by alxelax — DISMISSED** [review]

**Review by HaavardRei — DISMISSED** [review]

**Comment by omkar3141** [comment]
Fixed documentation build.

**Review by HaavardRei — APPROVED** [review]

**Review by alxelax — APPROVED** [review]

**Comment by alxelax** [comment]
Hi @carlescufi, this is ready for merge.

**Review by Balaklaka — APPROVED** [review]

**Pull request merged**
